libff/algebra/curves/bls12_381/bls12_381_g1.cpp

```diff
@@ -0,0 +1,525 @@
+
+bool bls12_381_G1::is_special() const
+{
+    return (this->is_zero() || this->Z == bls12_381_Fq::one());
+}
+
+bool bls12_381_G1::is_zero() const { return (this->Z.is_zero()); }
+
+bool bls12_381_G1::operator==(const bls12_381_G1 &other) const
+{
+    if (this->is_zero()) {
+        return other.is_zero();
+    }
+
+    if (other.is_zero()) {
+        return false;
+    }
+
+    /* now neither is O */
+
+    // using Jacobian coordinates so:
+    // (X1:Y1:Z1) = (X2:Y2:Z2)
+    // iff
+    // X1/Z1^2 == X2/Z2^2 and Y1/Z1^3 == Y2/Z2^3
+    // iff
+    // X1 * Z2^2 == X2 * Z1^2 and Y1 * Z2^3 == Y2 * Z1^3
+
+    bls12_381_Fq Z1_squared = (this->Z).squared();
+    bls12_381_Fq Z2_squared = (other.Z).squared();
+
+    if ((this->X * Z2_squared) != (other.X * Z1_squared)) {
+        return false;
+    }
+
+    bls12_381_Fq Z1_cubed = (this->Z) * Z1_squared;
+    bls12_381_Fq Z2_cubed = (other.Z) * Z2_squared;
+
+    return !((this->Y * Z2_cubed) != (other.Y * Z1_cubed));
+}
+
+bool bls12_381_G1::operator!=(const bls12_381_G1 &other) const
+{
+    return !(operator==(other));
+}
+
+bls12_381_G1 bls12_381_G1::operator+(const bls12_381_G1 &other) const
+{
+    // handle special cases having to do with O
+    if (this->is_zero()) {
+        return other;
+    }
+
+    if (other.is_zero()) {
+        return *this;
+    }
+
+    // http://www.hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-0.html#addition-add-2007-bl
+    // no need to handle points of order 2,4
+    // (they cannot exist in a prime-order subgroup)
+
+    // check for doubling case
+
+    // using Jacobian coordinates so:
+    // (X1:Y1:Z1) = (X2:Y2:Z2)
+    // iff
+    // X1/Z1^2 == X2/Z2^2 and Y1/Z1^3 == Y2/Z2^3
+    // iff
+    // X1 * Z2^2 == X2 * Z1^2 and Y1 * Z2^3 == Y2 * Z1^3
+
+    bls12_381_Fq Z1Z1 = (this->Z).squared();
+    bls12_381_Fq Z2Z2 = (other.Z).squared();
+
+    bls12_381_Fq U1 = this->X * Z2Z2;
+    bls12_381_Fq U2 = other.X * Z1Z1;
+
+    bls12_381_Fq Z1_cubed = (this->Z) * Z1Z1;
+    bls12_381_Fq Z2_cubed = (other.Z) * Z2Z2;
+
+    bls12_381_Fq S1 = (this->Y) * Z2_cubed; // S1 = Y1 * Z2 * Z2Z2
+    bls12_381_Fq S2 = (other.Y) * Z1_cubed; // S2 = Y2 * Z1 * Z1Z1
+
+    if (U1 == U2 && S1 == S2) {
+        // dbl case; nothing of above can be reused
+        return this->dbl();
+    }
+
+    // rest of add case
+    bls12_381_Fq H = U2 - U1; // H = U2-U1
+    bls12_381_Fq S2_minus_S1 = S2 - S1;
+    bls12_381_Fq I = (H + H).squared();          // I = (2 * H)^2
+    bls12_381_Fq J = H * I;                      // J = H * I
+    bls12_381_Fq r = S2_minus_S1 + S2_minus_S1;  // r = 2 * (S2-S1)
+    bls12_381_Fq V = U1 * I;                     // V = U1 * I
+    bls12_381_Fq X3 = r.squared() - J - (V + V); // X3 = r^2 - J - 2 * V
+    bls12_381_Fq S1_J = S1 * J;
+    bls12_381_Fq Y3 = r * (V - X3) - (S1_J + S1_J); // Y3 = r * (V-X3)-2 S1 J
+    bls12_381_Fq Z3 = ((this->Z + other.Z).squared() - Z1Z1 - Z2Z2) *
+                      H; // Z3 = ((Z1+Z2)^2-Z1Z1-Z2Z2) * H
```

Comment:
Here is an example of what I'm saying above re: the comments at end of line trigerring line breaks.
It makes the code harder to read. So let's move comments on the preceding line.

Reply:
Done.